rustls/src/server/mod.rs

```diff
@@ -146,6 +146,311 @@ impl<'a> ClientHello<'a> {
+/// ```
+///
+/// This may be shortened to:
+/// 
+/// ```ignore
+/// ServerConfigBuilder::defaults()
+///     .without_client_auth()
+///     .with_single_cert(certs, private_key)
+///     .build();
+/// ```
+///
+/// The types used here fit together like this:
+///
+/// 1. Get a ServerConfigBuilder with `ServerConfigBuilder::new()`.
+///    You must make a decision on which ciphersuites to use, typically
+///    by calling `with_default_ciphersuites()`.
+/// 2. You now have a ServerConfigBuilderWithSuites.  You must make a decision
+///    on key exchange groups: typically by calling `with_default_kx_groups()`.
+/// 2. You now have a ServerConfigBuilderWithSuitesAndKx.  You must make
+///    a decision on how and whether to use client authentication.  Perhaps
+///    you call `without_client_auth()`.
+/// 3. You now have a ServerConfigBuilderComplete.  This object has a number
+///    of defaults you can change, or you can pass it to `ServerConfig::new`.
+pub struct ServerConfigBuilder {}
+
+impl ServerConfigBuilder {
+    /// Start building a ServerConfig.
+    pub fn new() -> ServerConfigBuilder {
+        ServerConfigBuilder {}
+    }
+
+    /// Start building a ServerConfig, and accept defaults for underlying
+    /// cryptography.
+    ///
+    /// These are safe defaults, useful for 99% of applications.
+    ///
+    /// With the returned object, you must still make decisions on
+    /// client authentication and provide server credentials -- rustls
+    /// can't provide useful defaults for these.
+    pub fn defaults() -> ServerConfigBuilderWithSuitesAndKxGroups {
+        ServerConfigBuilder::new()
+            .with_default_ciphersuites()
+            .with_default_kx_groups()
+    }
+
+    /// Choose a specific set of ciphersuites.
+    pub fn with_ciphersuites(
+        self,
+        ciphersuites: &[&'static SupportedCipherSuite],
+    ) -> ServerConfigBuilderWithSuites {
+        ServerConfigBuilderWithSuites {
+            ciphersuites: ciphersuites.to_vec(),
+        }
+    }
+
+    /// Choose the default set of ciphersuites.
+    ///
+    /// Note that this default provides only high-quality suites: there is no need
+    /// to filter out low-, export- or NULL-strength ciphersuites: rustls does not
+    /// implement these.
+    pub fn with_default_ciphersuites(self) -> ServerConfigBuilderWithSuites {
+        self.with_ciphersuites(DEFAULT_CIPHERSUITES)
+    }
+}
+
+/// A ServerConfigBuilder where we know the ciphersuites.
+pub struct ServerConfigBuilderWithSuites {
+    ciphersuites: Vec<&'static SupportedCipherSuite>,
+}
+
+impl ServerConfigBuilderWithSuites {
+    /// Choose a specific set of key exchange groups.
+    pub fn with_kx_groups(
+        self,
+        kx_groups: &[&'static SupportedKxGroup],
+    ) -> ServerConfigBuilderWithSuitesAndKxGroups {
+        ServerConfigBuilderWithSuitesAndKxGroups {
+            ciphersuites: self.ciphersuites,
+            kx_groups: kx_groups.to_vec(),
+        }
+    }
```

Comment:
As far as I can tell, rustls did not even make key exchange groups configurable in the latest published release. Why is this necessary/useful?

Reply:
I agree that the key exchange groups should be configurable. Why not? If you're using a profile of TLS that, say, only uses P-256, or mandates P-256 be supported as a key exchange algorithm, then you need to be able to switch the initial client hello to use P-256 instead of X25519 or else you're going to suffer unnecessarily terrible performance.

Although this functionality wasn't in the last release, it is already in main; there was an issue filed about X25519 being hard-coded and a PR was merged to address it.

Reply:
Fair enough.